2단계 미션 제출합니다

iimpala opened the pull request.

Hue9010 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -3,13 +3,12 @@
+import java.util.Objects;
 import java.util.stream.Collectors;
 
 public class LottoTickets {
```
```java
  private int getRank(int matchedCount, boolean matchedBounus) {
    if (matchedCount == LottoTicket.LOTTO_NUMBERS_LENGTH) {
      return 1;
    }

    if (matchedCount == LottoTicket.LOTTO_NUMBERS_LENGTH - 1 && matchedBounus) {
      return 2;
    }

    if (matchedCount > 2) {
      return LottoTicket.LOTTO_NUMBERS_LENGTH - matchedCount + 2;
    }

    return 0;
  }
```

이 부분 숫자 값이아니라 LottoRank에서 matchedCount와 matchBounus를 통해 알맞은 LottoRank를 찾게 개선 할 수 있을 것 같네요.

Hue9010 reviewed (changes requested):
잘해주셨네요. 👍
머지하기엔 몇가지 아쉬운 부분이 있어서 코멘트 남겼어요. 
확인하시고 궁금한 부분 있으면 질문주세요! 😉

sumniy commented on src/main/java/controller/LottoController.java:
```diff
@@ -4,32 +4,39 @@
 
-  private static LottoTickets createLottoTickets(int ticketCount) {
-    LottoTickets lottoTickets = LottoService.generateRandomLottoTickets(ticketCount);
+  private static LottoTickets createLottoTickets(LottoTicketCount lottoTicketCount) {
+    List<String> manualLottoNumbers = LottoInputView
+        .inputManualLottoNumbers(lottoTicketCount.getManualCount());
+    LottoTickets lottoTickets = LottoService
+        .generateLottoTickets(manualLottoNumbers, lottoTicketCount.getAutoCount());
     LottoOutputView.printTickets(lottoTickets.getLottoTicketsInfo());
     return lottoTickets;
   }
```
Controller는 로또가 진행되는 흐름을 책임지게하고 거기서 로또 티켓을 생성하는 부분들을 Service로 분리했습니다. 
수동, 자동 로또 생성에 대한 로직을 따로 관리하려고 이렇게 했는데, 이게 Service의 역할에는 맞지 않는 것일까요?

sumniy replied:
다시 생각해보니 로또 티켓 생성은 Service로 생각하기보다는 컨트롤러가 가지거나 LottoTicketGenerator와 같이 관리하는게 맞는 것 같기도 합니다 🤔

Hue9010 replied:
사실 가장 큰 핵심은 Service의 명칭이 약간은 모호합니다. Spring의 Service layer를 의식한 네이밍이 아닐까 생각되네요. 그래서 Service가 로또의 생성에 대해서만 관리하는 걸 고려한건지 그 이상의 의미를 담고 있는지 문의 드렸던 겁니다. 의견 주신 것 처럼 LottoService보다는 LottoTicketGenerator가 좋을 것 같습니다. 👍 

제가 드린 예시의 상황은 Service에서 담당을 수행을 해주고 있으니 고려치 말아 주세요. 네이밍만 생각하면 될 것 같습니다.😭 🙏  `List<String> -> List<LottoNumber>`이 여기서 이뤄진다고 잘못 봤네요. 😱

sumniy replied:
아하 알겠습니다 🤗

sumniy commented:
피드백 반영했습니다. 리뷰 부탁드립니다. 🙇‍♂️

Hue9010 commented on src/main/java/domain/LottoRank.java:
```diff
@@ -2,35 +2,47 @@
 
-  private final int rank;
-  private final Function<Long, Long> expression;
+  private final int matchCount;
```
matchCount  👍

Hue9010 reviewed (approved):
피드백 잘 반영해주셨네요. 👍
이번 미션은 여기서 마무리하도록 할게요.
추가로 궁금한 사항 있으면 언제든 질문해주세요.
고생하셨습니다. 다음과제도 화이팅입니다!

sumniy commented:
감사합니다!

chanhalee committed:
docs(README) Todo List 작성

Todo List 작성

chanhalee committed:
feat(Budget) 구매 금액에 대한 요구사항 Test 작성 및 구현

구매 금액에 대한 요구사항 Test 작성 및 구현

chanhalee committed:
feat(Ball) 로또 번호에 대한 요구사항 test 작성 및 구현

로또 번호에 대한 요구사항 test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 정수여야 한다.

chanhalee committed:
feat(Ticket) 티켓에 대한 요구사항 Test 작성 및 구현

티켓에 대한 요구사항 Test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 중복되지 않는 6개의 수로 이루어져 있다

chanhalee committed:
refactor(Budget) validation 메서드 static 속성 제거

validation 메서드 static 속성 제거

chanhalee committed:
feat(WinningNumbers) 티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현

티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현
- 등수에 대한 Enum 클래스 Prize 작성
- Ball -> Number로 클래스 명 변경
- Number Set을 보유하는 Numbers 클래스 생성
- WinningNumbers, TicketNumbers가 Numbers를 상속 받도록 구현

chanhalee committed:
feat(PrizeCalculator) 수익률계산기 테스트 작성 및 구현

수익률계산기 테스트 작성 및 구현

chanhalee committed:
feat(LottoMachine) LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

chanhalee committed:
feat(LottoMachine) LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

chanhalee committed:
refactor MVC 패턴에 맞게 패키지 분리

chanhalee committed:
feat(LottoView) view 레이어 구현

view 레이어 구현

chanhalee committed:
feat(LottoController) Controller 구현

Controller 구현

iimpala committed:
refactor(dto) dto 패키지 분리

iimpala committed:
refactor(LottoNumber) 클래스 이름 변경

도메인의 의미를 좀 더 명확하게 나타낼 수 있도록 클래스 이름 변경
- Number -> LottoNumber

iimpala committed:
refactor(LottoTicket) 클래스명 변경

도메인의 의미를 좀 더 명확하게 나타낼 수 있도록 클래스 이름 변경
- TicketNumbers -> LottoTicket

iimpala committed:
refactor(WinningNumbers) WinningNumbers와 Numbers의 상속관계 제거

WinningNumbers가 Numbers를 상속받는 구조에서 WinningNumbers가 LottoTicket을 가지는 구조로 변경

iimpala committed:
refactor(LottoTicket) LottoTicket과 Numbers의 상속관계 제거

불필요한 상속관계를 제거하고 Numbers 클래스를 삭제

iimpala committed:
refactor(all) 상수 추출

도메인 객체들에서 상수 추출

iimpala committed:
refactor(LottoNumber) lottoNumberPool 제거

- 1 ~ 45까지의 LottoNumber를 캐싱하여 불칠요한 인스턴스를 생성하지 않도록 변경
- LottoNumber 생성자를 private로 변경
- LottoNumber에 getValue(), getValues(), valueOf() 정적 메소드 추가
- LottoTicket 생성방식 변경(numberPool 제거)

iimpala committed:
refactor(WinningLotto) 클래스면 변경

도메인간 포함관계와 성격을 고려하여 WinningNumbers -> WinningLotto로 클래스명 변경

iimpala committed:
refactor(LottoTicketSeller) 클래스명 변경 및 책임 분리

- 로또 판매와 결과확인 두가지 역할을 담당하던 LottoMachine에서 결과 확인 로직을 WinningLotto에 위임하여 한가지 책임만 담당하도록 변경
- 이에 따라 LottoMachine -> LottoTicketSeller로 클래스명 변경

iimpala committed:
refactor(Prize) getPrize 메소드 위치 변경

당첨 등수를 판별하는 로직을 Prize 클래스로 위임하여 객체의 응집도를 높임
- 메소드 이름 변경(getPrize -> evaluate)
- 테스트 코드 추가(PrizeTest)

iimpala committed:
refactor(LottoResult) LottoResult 관련 구조 리팩토링

- 도메인의 의미를 명확히 하기 위헤 클래스명 변경(Budget -> LottoPurchaseBudget, PrizeCalculator -> LottoResult)
- 수익률 계산 로직을 PrizeCalculator에서 LottoPurchaseBudget로 위임
- WinningLotto에서 단일 Ticket의 Prize를 반환하는 calculatePrize 메소드 작성
- 결과 집계를 WinningLotto에서 LottoResult로 위임
- 리팩토링과 클래스 추가로 인한 테스트 추가 및 변경

iimpala committed:
refactor(dto) 도메인에서 toDto 메소드 분리

view 레이어에서 도메인의 Prize를 알고있는 문제를 해결하기 위해 controller에서 도메인과 dto를 변환하는 작업을 수행
- Prize에서 toString 오버라이딩
- dto 패키지 위치 이동
- dto에서 Prize 의존성 제거

iimpala committed:
refactor(exception) 에외 타입 변경

validation 실패시 던지는 예외 타입을 RuntimeException -> IllegalArgumentException으로 변경

iimpala committed:
feat(LottoPurchaseBudget) 로또 비용차감 로직 작성

- 구입하려는 로또 개수를 받아 예산 안에 드는지 검증하는 로직 작성
- 로또 구입시 금액만큼 예산을 차감하는 로직 작성

iimpala committed:
feat(LottoTicketSeller) 수동 로또 구매 로직 작성

- 수동으로 로또 번호를 받아 구입하는 로직 추가
